Dataverse-specific support

Hwesta opened the pull request:
This includes #346 

Add support for parsing an incoming METS file generated from Dataverse, and updating the AIP METS based on this.
- Display dataset name as the display string
- Parse and validate Dataverse METS fileGrp checksums, add Event for Dataverse fixity
- Create derivations for Dataverse-generated .tab files
- Add Dataverse as an Agents based on information in the newly-created `agents.json`
- Set fileGrpUse's from the METS
- Copy dmdSecs to the AIP METS

See also: https://wiki.archivematica.org/Dataverse

Todo:
- [ ] Only try to parse the Dataverse METS if it's a Dataverse transfer type (or equivalent solution)

Hwesta committed:
API: display verbose name in browse display string

Hwesta committed:
Dataverse: add fixtures

Hwesta committed:
Dataverse: Add original files fixtures

Hwesta committed:
Dataverse: Parse incoming METS

Parse incoming METS to populate DB and validate transfer.

Hwesta committed:
Dataverse: Add checksum verification

Hwesta committed:
Dataverse: Add parse as MSCL TEMPORARY

Hwesta committed:
Dataverse: Parse agent info

Hwesta committed:
CreateMETS2: Add fileGrpUse "derivative"

Hwesta committed:
CreateMETS2: Parse Dataverse METS dmdSecs

Look for a metadata METS and copy the dmdSecs into the AIP METS.

Hwesta committed:
DB: Events and Agents are many-to-many

Database update only

Hwesta committed:
DatabaseFunctions: insertIntoEvents takes a list of Agents

Update for Events and Agents being many-to-many

Hwesta committed:
METS: Update for Event/Agents many-to-many relationship

Events only link to Agents explicitly associated with them. Agents are only
added to an amdSec if they're associated with Events in that amdSec.

createDigivProvMD creates both Events and Agents, as described. Created a new
createAgent that mimics createEvent in functionality. Update callers.

Add & update tests, update fixtures to be more modular.

Hwesta committed:
Models: Add UserProfile to link to Agent

Add a UserProfile extension to User for additional data, in this case linking
to an associated Agent. Fix primary key field types to what Django expects before
syncdb is run so Django can create the m2m table.

Hwesta committed:
Models: Added post-save signal for User->Agent creation

When a User is saved, create or update an Agent for that user, along with the
UserProfile. Add logging for models & Agent string representation.

Hwesta committed:
MCPServer: Update activeAgent to Agent ID not User ID

When a choice is made by a user, store the activeAgent associated with that
unit as the Agent ID, instead of the user's ID. Update logs to make unitvar
changes clearer.

Hwesta committed:
Install: Add pre-Django migrate hook

mistydemeo commented:
Looks like there are merge conflicts, rebase this?

mistydemeo commented on src/MCPClient/lib/clientScripts/archivematicaCreateMETS2.py:
```diff
@@ -807,6 +766,13 @@ def createFileSec(directoryPath, parentDiv, baseDirectoryPath, baseDirectoryName
+            if f.originallocation.endswith('.tab'):
+                dv_metadata = create_dataverse_tabfile_dmdsec(baseDirectoryPath, os.path.basename(f.originallocation))
+                dmdSecs.extend(dv_metadata)
+                ids = ' '.join([ds.get('ID') for ds in dv_metadata])
```
Nitpick, but you don't need the brackets when the comprehension is the only argument to a method call.

mistydemeo commented on src/MCPClient/lib/clientScripts/parse_dataverse.py:
```diff
@@ -0,0 +1,184 @@
+        with open(agents_jsonfile, 'r') as f:
+            agents_json = json.load(f)
+    except (OSError, IOError):
+        return None
```
Nitpick, but could be `return`

Hwesta replied:
Being explicit that it's not returning an ID but None instead.

jhsimpson committed:
adding default values for config parameters.

WIP - this is an example only.  This is a way to make sure there
are default values for a development/testing environment.
The assumption is that the user will override these with their own
settings, in a config file, but if the config file is missing or
is missing required elements, the defaults defined here will be used.

Suggesting this in response to
https://github.com/artefactual-labs/ansible-role-archivematica-src/pull/53

When deploying Archivematica using that ansible role, a new config file is
used, that is not compatible with this branch of the Archivematica repo.

By specifying default values, A dev install can be completed using ansible.

Hwesta committed:
Dataverse: update SQL changes

Hwesta committed:
Dataverse: ignore dataset.json in metadata dir

Hwesta committed:
Dataverse: Temporary: Ignore RData files

Ignore RData files since they aren't returned.

Hwesta commented:
Context: This PR was part of some experimental work for starting transfers from Dataverse.  The associated storage service PR artefactual/archivematica-storage-service#99 was merged because it was fairly standalone.  This also relies on files created by artefactual/automation-tools#15 to work properly.

This adds a microservice that parses the METS file generated by automation tools containing the Dataverse metadata.  However, if the Dataverse METS file does not exist it throws an error for all Standard transfers.  Before this can be merged, it needs to be updated to not require an external METS file.  This could be done through a new Dataverse transfer type, or by adding optional parsing of external metadata files.

A Dataverse transfer type is not preferred, since we're trying to reduce the number of transfer types overall, as they are error prone.  However, if we could autodetect a transfer type instead of requiring the user to provide it, that would make this option much more feasible.

Adding a generic parsing of external metadata files would also benefit DSpace, and make it easier to support other systems that want to pass in metadata with a transfer. 

Work to make Events and Agents have an explicit many-to-many relationship also came out of this work, and was merged as #346.  It is included in this PR, but on rebase only commits from "API: display verbose name in browse display string" onwards are needed. 

Remaining todos:
- [ ] Optionally parse the Dataverse METS file in `metadata/METS.xml`
- [ ] Convert SQL migration to Django migration
- [ ] Confirm the original RData files are being returned from Dataverse and remove "Dataverse: Temporary: Ignore RData files" commit
- [ ] Ask on the dataverse mailing list about the JSON structure, and what fields are required/optional

sevein commented:
Replaced by #1242.

The branch won't be maintained for future reference.
